lib/experimental/html/class-wp-html-tag-processor.php

```diff
@@ -1501,6 +1534,18 @@ public function is_tag_closer() {
 		return $this->is_closing_tag;
 	}
 
+	/**
+	 * Add a lexical update, i.e. a replacement of HTML at a given position.
+	 *
+	 * @param int    $start    The start offset of the replacement.
+	 * @param int    $end      The end offset of the replacement.
+	 * @param string $text  The replacement.
+	 * @return void
+	 */
+	protected function add_lexical_update( $start, $end, $text ) {
```

Comment:
We _probably_ need to flush updates here to avoid collisions.

Reply:
> We probably need to flush updates here to avoid collisions.

I'd really like to avoid any possibility of adding colliding updates here. I think at that point we've exposed too much of the internal logic of the system or we have failed to maintain our internal consistency.

This is something I've worried about as we look at making sure `get_attribute()` returns the updated values for enqueued `set_attribute()` calls.

Put another way, the fact that we are raising these concerns might be more evidence that we're trying to stuff too much in when we could call `get_updated_html()` before some of these operations and then not worry about it.

Reply:
Maybe I was being too pessimistic here. The example(s) I was thinking of all involved moving to a different tag (via `next_tag()` or `seek()`), but I had forgotten that we flush in those cases anyway, since our updates are per-tag.

Is it really that simple? Does the updating mechanism Just Work™️, even for more generic changes? 🤔

Reply:
> Does the updating mechanism Just Work™️, even for more generic changes?

Right now attribute updates have a name corresponding to the attribute name, so successive attribute updates overwrite previous ones. Other textual updates are appended without a name (so probably get some generic integer key).

If we add overlapping textual updates we could collide. If we add successive attribute updates they replace the earlier ones and things are safe.

On this note we might be able to run a variant of `parse_next_attribute()` on `get_attribute()` if we have updates enqueued for the given attribute. That could save us the complexity of tracking attribute updates separately in #46680

Reply:
> Right now attribute updates have a name corresponding to the attribute name, so successive attribute updates overwrite previous ones.

Yeah. With the changes in this PR, we'd be retaining this behavior (which is happening entirely inside `set_attribute`). Eventually, when they're carried over to `$lexical_updates` by `attribute_updates_to_lexical_updates`, those keys are dismissed.

> Other textual updates are appended without a name (so probably get some generic integer key).

Yeah -- I'm implementing this behavior more explicitly in this PR, where other textual updates go straight to `$lexical_updates`, which is a non-associative array.

> If we add overlapping textual updates we could collide. If we add successive attribute updates they replace the earlier ones and things are safe.

I'm starting to think that things really might work in our favor -- at least as long as we only support `set_content_inside_balanced_tags`: By its nature, that's an operation that shouldn't collide with `set_attribute` when applied to the same tag, e.g. the `section` in the following:

```
<section id="foobar">
   <div>Text</div>
</section>
```

Right now I'm thinking to maybe just throw a ton of test coverage at the problem with every possible scenario I can think of, to see if I'm missing something 😅 

---

One thing I'm pretty sure is that we _will_ need to invalidate (i.e. release?) bookmarks. Anything that points to a tag that's overwritten by `set_content_inside_balanced_tags` is going to get lost.

Reply:
I'd remove this method –  `$this->lexical_updates[] = new WP_HTML_Text_Replacement( $start, $end, $text );` can be done in-place where its needed.

> I'd really like to avoid any possibility of adding colliding updates here. I think at that point we've exposed too much of the internal logic of the system or we have failed to maintain our internal consistency.

+1 to that.

Reply:
> I'd remove this method – `$this->lexical_updates[] = new WP_HTML_Text_Replacement( $start, $end, $text );` can be done in-place where its needed.

I'm exposing it (`protected`) [for use by `set_content_inside_bookmarks`](https://github.com/WordPress/gutenberg/pull/47036/files#diff-1b6fe2a8597f91969a7d67aec6721b99e645db0f69ea8eca19b00e083aaeefc5R183) (which is a method of `WP_HTML_Processor`, which in turn is derived from `WP_HTML_Tag_Processor`). I don't mind removing the method, but it means I'll have to make `$lexical_updates` `protected`.

Reply:
05bc7fe7fa